File: src/csharp/Foundgine.Core/Semantic/AliasWeightEvidenceGate.cs
using Foundgine.Core.Abstractions;
using Foundgine.Core.Semantic.Resolution;

namespace Foundgine.Core.Semantic;

/// <summary>Describes whether weighted alias evidence was applicable to a resolution.</summary>
public enum AliasEvidenceStatus : byte
{
    /// <summary>No lexical alias with a declared weight participated in the resolution.</summary>
    NotApplicable,

    /// <summary>Weighted alias evidence participated and all observed identities met the threshold.</summary>
    Sufficient,

    /// <summary>Weighted alias evidence participated and at least one observed identity was below the threshold.</summary>
    Insufficient
}

/// <summary>
/// Whether the semantic model itself was already known with certainty prior to
/// lexical resolution. This is a different epistemic category from declared
/// alias evidence: it describes provenance of the model, not the strength of a
/// lexical match, and it deliberately has no numeric scale so it can never be
/// combined arithmetically (e.g. max/averaged) with an alias weight.
/// </summary>
public enum ModelResolutionEvidence : byte
{
    /// <summary>The model's identity was not asserted as already known.</summary>
    Unknown,

    /// <summary>The model's identity was known with certainty independent of lexical alias matching.</summary>
    KnownWithCertainty
}

/// <summary>
/// Measured, application-declared lexical evidence. This type does not perform
/// authorization and does not alter retrieval scores.
/// </summary>
public sealed record AliasWeightEvidenceResult(
    AliasEvidenceStatus Status,
    ModelResolutionEvidence ModelEvidence,
    IReadOnlyDictionary<EntityId, int> EntityWeights,
    IReadOnlyDictionary<FieldId, int> FieldWeights,
    IReadOnlyDictionary<RelationshipId, int> RelationshipWeights,
    IReadOnlyList<EntityId> ViolatingEntities,
    IReadOnlyList<FieldId> ViolatingFields,
    IReadOnlyList<RelationshipId> ViolatingRelationships,
    string ContractFingerprint)
{
    /// <summary>
    /// Compatibility projection. Prefer <see cref="Status"/> because
    /// NotApplicable and Sufficient are intentionally different states.
    /// </summary>
    [Obsolete("Use Status. IsConclusive is true for both NotApplicable and Sufficient.", false)]
    public bool IsConclusive => Status != AliasEvidenceStatus.Insufficient;

    /// <summary>
    /// Compatibility projection onto the old numeric representation.
    /// </summary>
    [Obsolete("Use ModelEvidence. Model provenance is not on the same numeric scale as declared alias weight.", false)]
    public int? ModelWeight => ModelEvidence == ModelResolutionEvidence.KnownWithCertainty ? 100 : null;
}

/// <summary>
/// Measures application-declared alias weights that actually participated in
/// lexical grounding. The resolver may use this measurement as a commitment
/// policy, but the evaluator itself never authorizes execution.
/// </summary>
public static class AliasWeightEvidenceGate
{
    public static AliasWeightEvidenceResult Evaluate(
        SemanticModel model,
        int minimumWeight,
        SemanticLexicalResolution? lexicalResolution = null,
        bool modelKnownWithCertainty = false)
    {
        ArgumentNullException.ThrowIfNull(model);
        model.EnsureFrozen();
        return Evaluate(model.CreateSnapshot(), minimumWeight, lexicalResolution, modelKnownWithCertainty);
    }

    public static AliasWeightEvidenceResult Evaluate(
        SemanticContractSnapshot model,
        int minimumWeight,
        SemanticLexicalResolution? lexicalResolution = null,
        bool modelKnownWithCertainty = false)
    {
        ArgumentNullException.ThrowIfNull(model);

        if (minimumWeight is < 1 or > 100)
            throw new ArgumentOutOfRangeException(nameof(minimumWeight), minimumWeight,
                "minimumWeight must be between 1 and 100 (inclusive).");

        var modelEvidence = modelKnownWithCertainty
            ? ModelResolutionEvidence.KnownWithCertainty
            : ModelResolutionEvidence.Unknown;

        if (lexicalResolution is null || lexicalResolution.Steps.Count == 0)
        {
            return new(
                AliasEvidenceStatus.NotApplicable,
                modelEvidence,
                new Dictionary<EntityId, int>(),
                new Dictionary<FieldId, int>(),
                new Dictionary<RelationshipId, int>(),
                [],
                [],
                [],
                model.ContractFingerprint);
        }

        var entityWeights = new Dictionary<EntityId, int>();
        var fieldWeights = new Dictionary<FieldId, int>();
        var relationshipWeights = new Dictionary<RelationshipId, int>();

        foreach (var step in lexicalResolution.Steps)
        {
            var candidate = step.Candidate;
            switch (candidate.Kind)
            {
                case SemanticLexicalCandidateKind.Entity:
                case SemanticLexicalCandidateKind.Node:
                    if (candidate.EntityId is EntityId entityId &&
                        model.TryGetAlias(entityId, candidate.Token, out var entityAliasWeight))
                        AccumulateMax(entityWeights, entityId, entityAliasWeight);
                    break;

                case SemanticLexicalCandidateKind.Field:
                    if (candidate.EntityId is EntityId fieldOwnerEntityId &&
                        candidate.FieldId is FieldId fieldId &&
                        model.TryGetAlias(fieldOwnerEntityId, fieldId, candidate.Token, out var fieldAliasWeight))
                        AccumulateMax(fieldWeights, fieldId, fieldAliasWeight);
                    break;

                case SemanticLexicalCandidateKind.Relationship:
                    if (candidate.RelationshipId is RelationshipId relationshipId &&
                        model.TryGetAlias(relationshipId, candidate.Token, out var relationshipAliasWeight))
                        AccumulateMax(relationshipWeights, relationshipId, relationshipAliasWeight);
                    break;

                case SemanticLexicalCandidateKind.Traversal:
                    // Traversals are derived graph paths, not declared aliases.
                    break;
            }
        }

        var violatingEntities = entityWeights.Where(x => x.Value < minimumWeight).Select(x => x.Key).ToArray();
        var violatingFields = fieldWeights.Where(x => x.Value < minimumWeight).Select(x => x.Key).ToArray();
        var violatingRelationships =
            relationshipWeights.Where(x => x.Value < minimumWeight).Select(x => x.Key).ToArray();

        var applicable = entityWeights.Count > 0 || fieldWeights.Count > 0 || relationshipWeights.Count > 0;
        var status = !applicable
            ? AliasEvidenceStatus.NotApplicable
            : violatingEntities.Length == 0 &&
              violatingFields.Length == 0 &&
              violatingRelationships.Length == 0
                ? AliasEvidenceStatus.Sufficient
                : AliasEvidenceStatus.Insufficient;

        return new(
            status,
            modelEvidence,
            entityWeights,
            fieldWeights,
            relationshipWeights,
            violatingEntities,
            violatingFields,
            violatingRelationships,
            model.ContractFingerprint);
    }

    private static void AccumulateMax<TKey>(Dictionary<TKey, int> values, TKey key, int weight)
        where TKey : notnull
    {
        if (!values.TryGetValue(key, out var existing) || weight > existing)
            values[key] = weight;
    }
}

/// <summary>
/// Backing evidence attached to one grounding interpretation. Alias weights
/// are integers on their declared 1-100 scale; interpretation scores remain
/// separate doubles because they are retrieval/graph ranking heuristics.
/// ContractFingerprint identifies the exact frozen semantic contract this
/// evidence was measured against, so an audit trail can prove which contract
/// version a Sufficient/Insufficient verdict corresponds to.
/// </summary>
public sealed record AliasInterpretationEvidence(
    AliasEvidenceStatus Status,
    IReadOnlyDictionary<EntityId, int> EntityWeights,
    IReadOnlyDictionary<FieldId, int> FieldWeights,
    IReadOnlyDictionary<RelationshipId, int> RelationshipWeights,
    string? ContractFingerprint = null)
{
    public static AliasInterpretationEvidence From(AliasWeightEvidenceResult result) =>
        new(result.Status, result.EntityWeights, result.FieldWeights, result.RelationshipWeights,
            result.ContractFingerprint);
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/AliasWeightEvidenceGate.java
package com.foundgine.core.semantic;

import com.foundgine.core.abstractions.*;
import com.foundgine.core.semantic.resolution.*;
import java.util.*;

/** Measures declared alias-weight evidence that actually participated in lexical grounding. */
public final class AliasWeightEvidenceGate {
    private AliasWeightEvidenceGate() {}

    public enum AliasEvidenceStatus { NOT_APPLICABLE, SUFFICIENT, INSUFFICIENT }
    public enum ModelResolutionEvidence { UNKNOWN, KNOWN_WITH_CERTAINTY }

    public record AliasWeightEvidenceResult(
        AliasEvidenceStatus status,
        ModelResolutionEvidence modelEvidence,
        Map<EntityId,Integer> entityWeights,
        Map<FieldId,Integer> fieldWeights,
        Map<RelationshipId,Integer> relationshipWeights,
        List<EntityId> violatingEntities,
        List<FieldId> violatingFields,
        List<RelationshipId> violatingRelationships,
        String contractFingerprint) {
        public AliasWeightEvidenceResult {
            entityWeights = entityWeights == null ? Map.of() : Map.copyOf(entityWeights);
            fieldWeights = fieldWeights == null ? Map.of() : Map.copyOf(fieldWeights);
            relationshipWeights = relationshipWeights == null ? Map.of() : Map.copyOf(relationshipWeights);
            violatingEntities = violatingEntities == null ? List.of() : List.copyOf(violatingEntities);
            violatingFields = violatingFields == null ? List.of() : List.copyOf(violatingFields);
            violatingRelationships = violatingRelationships == null ? List.of() : List.copyOf(violatingRelationships);
        }
        /** Compatibility projection retained for parity with the C# obsolete API. */
        @Deprecated public boolean isConclusive() { return status != AliasEvidenceStatus.INSUFFICIENT; }
        /** Compatibility projection; model provenance is not a lexical score. */
        @Deprecated public Integer modelWeight() { return modelEvidence == ModelResolutionEvidence.KNOWN_WITH_CERTAINTY ? 100 : null; }
    }

    public static AliasWeightEvidenceResult evaluate(SemanticModel model, int minimumWeight,
                                                      SemanticLexicalResolution lexicalResolution,
                                                      boolean modelKnownWithCertainty) {
        Objects.requireNonNull(model); model.ensureFrozen();
        return evaluate(model.createSnapshot(), minimumWeight, lexicalResolution, modelKnownWithCertainty);
    }
    public static AliasWeightEvidenceResult evaluate(SemanticModel model, int minimumWeight) {
        return evaluate(model, minimumWeight, null, false);
    }
    public static AliasWeightEvidenceResult evaluate(SemanticContractSnapshot model, int minimumWeight,
                                                      SemanticLexicalResolution lexicalResolution,
                                                      boolean modelKnownWithCertainty) {
        Objects.requireNonNull(model);
        if (minimumWeight < 1 || minimumWeight > 100) throw new IllegalArgumentException("minimumWeight must be between 1 and 100 (inclusive).");
        var modelEvidence = modelKnownWithCertainty ? ModelResolutionEvidence.KNOWN_WITH_CERTAINTY : ModelResolutionEvidence.UNKNOWN;
        if (lexicalResolution == null || lexicalResolution.steps().isEmpty())
            return new AliasWeightEvidenceResult(AliasEvidenceStatus.NOT_APPLICABLE, modelEvidence, Map.of(), Map.of(), Map.of(), List.of(), List.of(), List.of(), model.contractFingerprint());

        var ew = new HashMap<EntityId,Integer>(); var fw = new HashMap<FieldId,Integer>(); var rw = new HashMap<RelationshipId,Integer>();
        for (var step : lexicalResolution.steps()) {
            var c = step.candidate();
            switch (c.kind()) {
                case ENTITY, NODE -> { if (c.entityId()!=null) model.aliasWeight(c.entityId(), c.token()).ifPresent(v -> ew.merge(c.entityId(), v, Math::max)); }
                case FIELD -> { if (c.entityId()!=null && c.fieldId()!=null) model.aliasWeight(c.entityId(), c.fieldId(), c.token()).ifPresent(v -> fw.merge(c.fieldId(), v, Math::max)); }
                case RELATIONSHIP -> { if (c.relationshipId()!=null) model.aliasWeight(c.relationshipId(), c.token()).ifPresent(v -> rw.merge(c.relationshipId(), v, Math::max)); }
                case TRAVERSAL, VALUE, OPERATION -> { }
            }
        }
        var ve = ew.entrySet().stream().filter(x->x.getValue()<minimumWeight).map(Map.Entry::getKey).toList();
        var vf = fw.entrySet().stream().filter(x->x.getValue()<minimumWeight).map(Map.Entry::getKey).toList();
        var vr = rw.entrySet().stream().filter(x->x.getValue()<minimumWeight).map(Map.Entry::getKey).toList();
        var applicable = !ew.isEmpty() || !fw.isEmpty() || !rw.isEmpty();
        var status = !applicable ? AliasEvidenceStatus.NOT_APPLICABLE : (ve.isEmpty() && vf.isEmpty() && vr.isEmpty() ? AliasEvidenceStatus.SUFFICIENT : AliasEvidenceStatus.INSUFFICIENT);
        return new AliasWeightEvidenceResult(status, modelEvidence, ew, fw, rw, ve, vf, vr, model.contractFingerprint());
    }
    public static AliasWeightEvidenceResult evaluate(SemanticContractSnapshot model, int minimumWeight) { return evaluate(model, minimumWeight, null, false); }

    public record AliasInterpretationEvidence(AliasEvidenceStatus status, Map<EntityId,Integer> entityWeights,
                                               Map<FieldId,Integer> fieldWeights, Map<RelationshipId,Integer> relationshipWeights,
                                               String contractFingerprint) {
        public static AliasInterpretationEvidence from(AliasWeightEvidenceResult result) {
            return new AliasInterpretationEvidence(result.status(), result.entityWeights(), result.fieldWeights(), result.relationshipWeights(), result.contractFingerprint());
        }
    }
}
